test/rulesets/Base/fastmath_able.jl

```diff
@@ -0,0 +1,107 @@
+# Add tests to the quote for functions with  FastMath varients.
+const FASTABLE_AST = quote
+    @testset "Trig" begin
+        @testset "Basics" for x = (Float64(π)-0.01, Complex(π, π/2))
+            test_scalar(sin, x)
+            test_scalar(cos, x)
+            test_scalar(tan, x)
+        end
+        @testset "Hyperbolic" for x = (Float64(π)-0.01, Complex(π-0.01, π/2))
+            test_scalar(sinh, x)
+            test_scalar(cosh, x)
+            test_scalar(tanh, x)
+        end
+        @testset "Inverses" for x = (0.5, Complex(0.5, 0.25))
+            test_scalar(asin, x)
+            test_scalar(acos, x)
+            test_scalar(atan, x)
+        end
+        @testset "Multivariate" begin
+            @testset "sincos" begin
+                x, Δx, x̄ = randn(3)
+                Δz = (randn(), randn())
+
+                frule_test(sincos, (x, Δx))
+                rrule_test(sincos, Δz, (x, x̄))
+            end
+        end
+    end  # Trig
+
+    @testset "exponents" begin
+        for x in (-0.1, 6.4)
+            test_scalar(inv, x)
+
+            test_scalar(exp, x)
+            test_scalar(exp2, x)
+            test_scalar(exp10, x)
+            test_scalar(expm1, x)
+
+            test_scalar(cbrt, x)
+
+            if x >= 0
+                test_scalar(sqrt, x)
+                test_scalar(log, x)
+                test_scalar(log2, x)
+                test_scalar(log10, x)
+                test_scalar(log1p, x)
+            end
+        end
+    end
+
+    @testset "Unary complex functions" begin
+        for x in (-4.1, 6.4)
+            test_scalar(abs, x)
+
+            test_scalar(angle, x)
+            test_scalar(abs2, x)
+            test_scalar(conj, x)
+
+        end
+    end
+
+    @testset "Unary functions" begin
+        for x in (-4.1, 6.4)
+            test_scalar(+, x)
+            test_scalar(-, x)
+        end
+    end
+
+    @testset "binary function ($f)" for f in (/, +, -, hypot, atan, rem, ^, max, min)
+        x, Δx, x̄ = 10rand(3)
+        y, Δy, ȳ = rand(3)
+        Δz = rand()
+
+        frule_test(f, (x, Δx), (y, Δy))
+        rrule_test(f, Δz, (x, x̄), (y, ȳ))
+    end
+
+
+
+    @testset "sign" begin
+        @testset "at points" for x in (-1.1, -1.1, 0.5, 100)
+            test_scalar(sign, x)
+        end
+
+        @testset "Zero over the point discontinuity" begin
+            # Can't do finite differencing because we are lying
```

Comment:
Curious, what do you mean by this?

Reply:
There is some math involve but basically the defintion of derivative is taking limit of small steps around a point.
It  is consider to be not defined if when you take those steps to the left and to the right you get a different answer.
In this case you get `lim x->0 +1/x` and `lim x-> 0 -1/x`
which we can say is something like positive and negativ infinity.

There is some extension of the math for deritivatives called the subgradient method.
and one of the outcomes of it is that it is good to treat derivatives that are not defined,
but that are such that from one side the limit is positive and from the other it is negative as being zero.
This is because one of the main uses of the dervative is to find local minima,
which typically are defined as being derivative zero at that point.
But for in cases where the dervative at that point is not define then it is also true that if the derivative on either side is opposite in sign it is also a local minima